File: src/main/java/cp/dojo/solution/arrays/ThreeSum.java
package cp.dojo.solution.arrays;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class ThreeSum {
  public static void main(String[] args) {
    ThreeSum s = new ThreeSum();
    List<List<Integer>> triplets = s.threeSum(new int[]{-1,0,1,2,-1,-4});
    for(List<Integer> tripl : triplets){
      System.out.println(tripl);
    }
  }

  public List<List<Integer>> threeSum(int[] nums) {
    //Sort nums
    Arrays.sort(nums);
    List<List<Integer>> triplets = new ArrayList<>();
    for(int i = 0; i < nums.length; ++i){
      //If i != 0 && nums[i] == nums[i-1] skip computation
      if(i != 0 && nums[i] == nums[i-1]){
        continue;
      }
      int low = i+1, high = nums.length-1;
      while(low < high){
        int sum = nums[i] + nums[low] + nums[high];
        if(sum == 0){
          triplets.add(List.of(nums[i], nums[low], nums[high]));
          ++low; --high;
          while(low < high && nums[low] == nums[low-1]){
            ++low;
          }
          while(low < high && nums[high] == nums[high+1]){
            --high;
          }
        } else if(sum > 0){
          --high;
        } else {
          ++low;
        }
      }
    }
    return triplets;
  }
}

File: arrays/ThreeSum.kt
package arrays

fun main() {
    for (triplets: List<Int> in threeSum(intArrayOf(-1, 0, 1, 2, -1, -4))) {
        println(triplets.joinToString())
    }

    for (triplets: List<Int> in threeSum(intArrayOf(1, -1, -1, 0))) {
        println(triplets.joinToString())
    }
}


fun threeSum(nums: IntArray): List<List<Int>> {
    nums.sort()
    val res = ArrayList<List<Int>>()
    for (i in 0 until nums.size) {
        if (i != 0 && nums[i] == nums[i - 1]) {
            continue
        }
        var j = i + 1;
        var k = nums.size - 1
        while (j < k) {
            val sum = nums[i] + nums[j] + nums[k]
            if (sum == 0) {
                res.add(listOf(nums[i], nums[j], nums[k]))
                ++j
                --k

                while (j < k && nums[j] == nums[j - 1]) ++j
                while (j < k && nums[k] == nums[k + 1]) --k
            } else if (sum > 0) --k
            else ++j
        }
    }
    return res
}